### PR TITLE
[WiP][CR]Vehicle traction

### PR DESCRIPTION
**Pull request by Coolthulhu**
Problems:
- Vehicles don't need wheel area, deathrollers can roll on regular car wheels just fine
- Size is king, high mass won't slow down a vehicle much because the above point is true
- Offroading is not a problem, the worst thing that can happen is collision with a shrub
- Center of mass doesn't mean much
- Amphibious vehicles don't exist

Solution:
- Make non-road surfaces provide a multiplier on vehicle's ability to convert torque to momentum. Scaled by wheel area and distance between center of mass and "center of support".
- Make non-road surfaces subtract a flat value from vehicle's current velocity, depending on its mass, velocity and mass' distribution over wheels. When balancing, assume that current vehicle calculations were all about vehicles on roads.
- Rewrite boats. Have each boat part provide some amount of maximum weight, if sum of those is below vehicle's weight, let it float, otherwise have it sink. Involve center of mass by scaling down that maximum weight heavily if CoM is not over weighted center of boat parts.

Would be useful if someone with vehicle physics knowledge could provide some relevant data.
How do cars behave on roads vs. off roads, how does that change with heavy load, how are trucks different, how easily can a car get clogged down in mud depending on its mass and wheel setup etc.
Also similar data for boats.

### CODE REVIEW

**Comment by mugling** [comment]
> Offroading is not a problem

This is particularly interesting if we enhance mapgen to generate turrets, barricades and such on the main roads

> Amphibious vehicles don't exist

Sounds interesting. Amphibious or submersible?

**Comment by Coolthulhu** [comment]
> Amphibious or submersible?

Amphibious. Submersibles would show lack of water z-levels too well.

**Comment by Coolthulhu** [comment]
...I don't really know how to handle the physics.
I think I'll cherrypick the weight caching and commit it separately because it works.

**Comment by illi-kun** [comment]
I did some investigations concerning traction, here is [a good source of data](https://translate.google.ru/translate?sl=ru&tl=en&js=y&prev=_t&hl=ru&ie=UTF-8&u=http%3A%2F%2Ftransporton.ru%2Fautotransport%2F149-tyagovo-skorostnye-svojstva-avtomobilya.html%3Fstart%3D8&edit-text=&act=url). Since auto-translation works ugly, I'll try to translate the most important things below. 

On the longitudinal friction coefficient φx have the influence on the many design and operating factors. It is determined experimentally. The values ​​of the average φx for different road conditions and the surfaces are presented below:

Type of Road | Dry Condition | Wet Condition
----|----|---
Asphalt road | 0.7 - 0.8 | 0.35 - 0.45
Gravel road | 0.6 - 0.7 | 0.3 - 0.4
Dirt road | 0.5 - 0.6 | 0.2 - 0 4
Snow | 0.2 | 0.3
Ice  | 0.1 | 0.2

**Comment by DeadLeaves** [comment]
@illi-kun I like that data. Would make winter a lot more scary due to slower vehicles.

If asphalt road is the standard, then scale up the 0.7-0.8 to 1. and scale the other values accordingly. (for easier overview)

**Comment by Zireael07** [comment]
I like the way you're thinking, guys!

**Comment by mugling** [comment]
> @illi-kun I like that data. Would make winter a lot more scary due to slower vehicles.

This is a good idea

**Comment by CIB** [comment]
Just some points, as I talked about on IRC:
- Non-road surfaces shouldn't reduce a vehicle's velocity, instead it should factor into "acceleration" code, requiring a constant energy expense to counter-act
- Acceleration code should also be updated to factor in how much torque / force is required to accelerate on this terrain at all. This should particularly cover accelerating the vehicle from a stand-still or low speed

I realize that this would require rewriting the code extensively, but I believe that if we try to fix this code at all, it would be better to do so in a way that it's consistent and makes sense as a whole, rather than patching it until it somehow resembles what we want.

**Comment by Coolthulhu** [comment]
> Non-road surfaces shouldn't reduce a vehicle's velocity, instead it should factor into "acceleration" code, requiring a constant energy expense to counter-act

I meant to do that too, the part about current velocity was mostly about vehicle not rolling too far when left with idle engine. On a road you could coast for a while, but on a muddy surface you'd probably stop after just few steps of coasting.

The biggest problem with acceleration I'll have to solve will probably not be passing the values around, but cruise control vs. safe speed.

**Comment by CIB** [comment]
> I meant to do that too, the part about current velocity was mostly about vehicle not rolling too far when left with idle engine. On a road you could coast for a while, but on a muddy surface you'd probably stop after just few steps of coasting.

Well, basically it'd be nice to have a central place in the code where all this stuff is addressed. So the code responsible for "acceleration" might decide that since the vehicle is no longer accelerating and there's friction, it's losing speed now.

**Comment by CIB** [comment]
In essence, I'm suggesting a function that looks a bit like this:

Inputs: Current vehicle, speed, mass, wheels, terrain below wheels, target speed
Intermediate result: Forces that counter-act acceleration (or would cause decelleration), maximum amount of force toward acceleration (result of maximum vehicle power and grip of wheels)
Output: New vehicle speed, vehicle power used, energy (fuel) expended

File: src/map.cpp [hunk 1]
```diff
@@ -605,69 +605,141 @@ bool map::vehicle_falling( vehicle &veh )
...
 }
 
-float map::vehicle_traction( vehicle &veh ) const
+float map::traction_mult( const tripoint &p ) const
 {
-    const tripoint pt = veh.global_pos3();
-    // TODO: Remove this and allow amphibious vehicles
-    if( !veh.floating.empty() ) {
-        return vehicle_buoyancy( veh );
+    const ter_t ter = ter_at( p );
```
Comment (BevapDin) [inline src/map.cpp]: Please don't copy the terrain data, a reference should work just fine. Same in the next function.

**Comment by Coolthulhu** [comment]
Closing for now, I'll need to rework it before it's mergeable.

**Pull request closed without merging**